### PR TITLE
Ups support

### PR DESCRIPTION
**Pull request by ajsalagundi1**

### CODE REVIEW

File: carriers/ups.js [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        activitiesList.location = geography.addressToString(activitiesList.address);
+    }
+
+    return activitiesList;
+}
+
+function filter(res) {
+    const packageInfo = res.body.TrackResponse.Shipment.Package;
+    var activitiesList = [];
+    if(packageInfo === undefined){
```
Comment (javamatte) [inline carriers/ups.js]: `if(` should be `if (` (needs a space)
I'm wondering if the ESLint rules for this project are missing/incorrect. There are a few lines (like this one) that should be throwing ESLint errors per our current style.
@freshlogic space-after-if and no-space-after-function seem to be missing (I made those names up), right?

File: carriers/ups.js [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        activitiesList.push(getActivities(res.body.TrackResponse.Shipment));
+    } else {
+        if (packageInfo.length === undefined) {
+            activitiesList.push(getActivities(packageInfo));
+        } else {
+            packageInfo.forEach((package) => {
+                activitiesList.push(getActivities(package));
+            })
```
Comment (javamatte) [inline carriers/ups.js]: missing ;

File: carriers/ups.js [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+                events: []
+            };
+
+            if (!res.body.TrackResponse) {
+                if (res && res.body && res.body.Fault && res.body.Fault.detail && res.body.Fault.detail.Errors && res.body.Fault.detail.Errors.ErrorDetail && res.body.Fault.detail.Errors.ErrorDetail.PrimaryErrorCode && res.body.Fault.detail.Errors.ErrorDetail.PrimaryErrorCode.Code === '151044') {
+                    // No Tracking Information
+                    return callback(null, results);
+                }
+
+                if (res && res.body && res.body.Fault && res.body.Fault.detail && res.body.Fault.detail.Errors && res.body.Fault.detail.Errors.ErrorDetail && res.body.Fault.detail.Errors.ErrorDetail.PrimaryErrorCode && res.body.Fault.detail.Errors.ErrorDetail.PrimaryErrorCode.Description) {
```
Comment (javamatte) [inline carriers/ups.js]: I can't think of one at the moment, but I wish there was a better way to do this that didn't involve so many expressions in the conditional. We do this sort of stuff often, but it's usually not so deep.
Not sure we need to change it, but I couldn't not say anything?

File: carriers/ups.js [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            var locations = [];
+
+            if (activitiesList[0].length === undefined) {
+                locations = Array.from(new Set(activitiesList.map(activity => activity.location)));
+                getResults(locations, callback, results, activitiesList);
+            } else {
+                activitiesList.forEach(activities => {
+                    locations = Array.from(new Set(activities.map(activity => activity.location)));
+                    getResults(locations, callback, results, activities);
+                })
```
Comment (javamatte) [inline carriers/ups.js]: ;

File: util/confirmUpsFreight.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+const checkDigit = require('../util/checkDigit')
+module.exports = function (trackingNumber) {
+    const firstChar = `${(trackingNumber.charCodeAt(0) - 63) % 10}`;
+    const remaining = trackingNumber.slice(1);
+    trackingNumber = `${firstChar}${remaining}`;
+    if (checkDigit(trackingNumber, [3, 1, 7], 10)) { return [true, true]; }
+    return [false, false];
+
+
```
Comment (javamatte) [inline util/confirmUpsFreight.js]: Run auto-format on this? (extra lines and 1-line if on line 6 will probably get auto-fixed)

File: carriers/ups.js [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+                    state: activity.ActivityLocation.Address === undefined ? (activity.ActivityLocation.StateProvinceCode === undefined ? '' : activity.ActivityLocation.StateProvinceCode) : (activity.ActivityLocation.Address.StateProvinceCode === undefined ? '' : activity.ActivityLocation.Address.StateProvinceCode),
+                    country: activity.ActivityLocation.Address === undefined ? (activity.ActivityLocation.CountryCode === undefined ? '' : activity.ActivityLocation.CountryCode) : (activity.ActivityLocation.Address.CountryCode === undefined ? '' : activity.ActivityLocation.Address.CountryCode),
+                    zipcode: activity.ActivityLocation.Address === undefined ? (activity.ActivityLocation.PostalCode === undefined ? '' : activity.ActivityLocation.PostalCode) : (activity.ActivityLocation.Address.PostalCode === undefined ? '' : activity.ActivityLocation.Address.PostalCode)
+                }
+                activity.location = geography.addressToString(activity.address);
+            } else {
+                activity.address = {
+                    city: '',
```
Comment (freshlogic) [inline carriers/ups.js]: Don't use empty string to represent null or undefined: https://coding.abel.nu/2013/01/on-null-or-why-empty-strings-are-not-same-as-null/

File: carriers/ups.js [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+function getActivities(package) {
+    var activitiesList = package.Activity;
+
+    if (activitiesList.length != undefined) {
+        activitiesList.forEach(activity => {
+            if (activity.ActivityLocation != undefined) {
+                activity.address = {
+                    city: activity.ActivityLocation.Address === undefined ? (activity.ActivityLocation.City === undefined ? '' : activity.ActivityLocation.City) : (activity.ActivityLocation.Address.City === undefined ? '' : activity.ActivityLocation.Address.City),
```
Comment (freshlogic) [inline carriers/ups.js]: This is very similar to the code on line 67 but doesn't look quite identical. This is usually the sign of a subtle bug. What is this code trying to accomplish?

File: carriers/ups.js [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    const checkdigit = (sum % 10) > 0 ? 10 - (sum % 10) : 0;
+    if (checkdigit === parseInt(trackingNumber[17], 10)) { return [true, true]; }
+    return [false, false];
+}
+
+function confirmUpsFreight(trackingNumber) {
+    const firstChar = `${(trackingNumber.charCodeAt(0) - 63) % 10}`;
+    const remaining = trackingNumber.slice(1);
+    trackingNumber = `${firstChar}${remaining}`;
+    if (checkDigit(trackingNumber, [3, 1, 7], 10)) { return [true, true]; }
```
Comment (freshlogic) [inline carriers/ups.js]: `return [true, true]`?

File: carriers/usps.js [hunk 9]
```diff
@@ -94,6 +94,8 @@ function USPS(options) {
 
                 const trackDetailList = data.TrackResponse.TrackInfo[0].TrackDetail;
 
+                //console.log(data.TrackResponse.TrackInfo[0].TrackDetail);
```
Comment (freshlogic) [inline carriers/usps.js]: Remove console.logs (even if they're comments)

File: package.json [hunk 10]
```diff
@@ -4,7 +4,7 @@
         "async": "~3.1.0",
         "moment-timezone": "~0.5.25",
         "node-geocoder": "~3.23.0",
-        "petty-cache": "~2.4.1",
+        "petty-cache": "^2.4.1",
```
Comment (freshlogic) [inline package.json]: Why change `~` to `^`?

**Pull request merged**